### DISCUSSION TITLE
Serverless adapter

### DISCUSSION THREAD

**Discussion by tzsk** [adonisjs/core#1256] [Ideas]
The serverless request context is different than normal node server request, because there's no node server in Serverless.
Which is why, if we could do a Serverless adapter for adonis ts that'll be a treat.
Do you have any plans for this in the future? Or recently.
Would be sick if I could run Afonis on a Lambda 🔥😍

**Comment by RomainLanz** [comment]
No, we have no plan doing that.

You are doing it wrong if you want to run a long-running process in Lambda, [what you need is PaaS, not FaaS](https://twitter.com/romainlanz/status/1283464773527580680).

**Reply by tzsk** [reply]
I'm not sure if I fully agree, JS is also a functional programming language. We just need to have a way to boot up the application with a Platform and execute depending on that.

Maybe we can have an Interface for `Platform` and replace the places we are accessing Platform Variables like `Node HTTP` with an implementation of that so that we can create different drivers for it. Maybe extract a new package `@adonisjs/platform-node` Someone might create their own platform like `adonis-platform-serverless` which only transforms the Serverless Context and create an implementation adhering to the Platform Interface. Then we can boot the app with the Platform driver.

I'm sorry if I went too far. But I don't like the `aws-serverless-express` and I'm trying to find a way to use adonis in our org.

**Reply by RomainLanz** [reply]
> I'm not sure if I fully agree, JS is also a functional programming language. We just need to have a way to boot up the application with a Platform and execute depending on that.

This is not the issue here. FaaS has been created to run **functions** in the cloud, it's generaly very fast and it scales them when needed (and shut them down when you don't need them).

It's really useful for short-running functions, like scripts. For example, you may need it to handle file uploads, image processing, sending mail, etc.

What you want to do is run a long-running process. The process is loaded in memory and stay there indefinitely. Booting-up the application everytime you hit your FaaS URI is extremely inefficient.

As said in my first message, there's a name for that and it's called PaaS. It runs your application in the cloud (like FaaS is doing) and keep your application alive for a specified time.

You can read the documentation of the [Google App Engine](https://cloud.google.com/appengine) if you want to try a PaaS solution.

**Reply by LudovicLemarinel** [reply]
Hi @RomainLanz 
It seems you don't know correctly AWS Lambda and how it works (No offence here !) : 

Sure , when your URI is hit the first time, you will experience a cold start : AWS is downloading your app ZIP from S3, then unzipping it, then run it, and pass the http context to your app.
BUT, then, it will stay warm, in memory and ready to receive more calls for about 15 minutes. 
While in those 15 minutes, your app is responding very quickly, as if it was a long running process .
And another cool effect, is that while in those 15 minutes, you are charged ONLY FOR ACTIVE CPU. So, if your endpoint is called once, you surely will billed for the coldstart time (about 300ms for a simple "hello world" Express project, less than 1.5s for a full Laravel project i.e....), but every next calls will be billed for the real duration (10 ms or so...)

So, even the AWS Lambda service was not designed to run Express apps, it's working, and it's used by many other projects (See Laravel VAPOR, which is fully backed by AWS Lambda !)

I personnaly use it for a big Laravel monolith, and it's so cool to not have to worry about servers being attacked, having to remember to update servers, upgrade them to major versions every 2 years....And it's really cheap : my Laravel monolith which handled 200k requests last month, only cost 0.20$ for the WHOLE MONTH (and it could be cheaper with node... its average response time is 200ms...) !

And cherry on the cake : it scales from 0 (you don't pay any cents when nobody uses your app) to 1000 instances if needed, without even having to worry about it, and without any extra cost :  If my FrontEnd make 10 calls in parallel, let's say to load my dashboard widgets, then AWS will run 10 lambdas in parallel, keep them warm for free (if ever I need to handle a bunch of queries...), and shut them down later if they are inactive...

So, to make it short, YES, it was not meant to run express / node http servers, but YES , it does run it, it is cheap and it is really convenient ;)

**Reply by RomainLanz** [reply]
I totally know how it works.

> So, even the AWS Lambda service was not designed to run Express apps, it's working, and it's used by many other projects (See Laravel VAPOR, which is fully backed by AWS Lambda !)

As you said, it is not designed to run a long-running process. I don't care if it works or not. Also, PHP is **not** a long-running process, so it is a great candidate for lambda.

I also don't care about spending 0$, 0.20$, 5$, or even 20$ per month for a server. Stop making things complicated for no reason.

**Reply by LudovicLemarinel** [reply]
You're not totally right : Laravel with OCTANE is a long running process... and that's how Laravel Vapor run php projects.

Having to manage servers and scale them up/down is what I see complicated for no reasons... 
Having to think before the users hit the server how big should it be is overcomplicated for no reasons.
Having to  migrate to a bigger server because we did scaled it too low is overcomplicated for no reasons...
We're on 2024, and we have services that do abstract those things for us... I really think serverless should be the correct way to host APIs today :)
And to add on top on it, we're pushed from all sides to do "green it"... Having to setup a big server "just in case..." instead of sharing CPU power among hundreds of customers is a non-sense !

As a freelance dev, I still smile when I hear my customers telling me "we're paying 200$ a month for our server, we have 4 users on it... how can you tell me it could cost only 0.20$ ??"...
Going from 200$/month to 0.2$/month for a small company can be a big game changer...

**Reply by RomainLanz** [reply]
Vapor does not run Laravel Octane by default. It is an opt-in option you can do in any Laravel application deployed anywhere.

> Ce're on 2024, and we have services that do abstract those things for us... I really think serverless should be the correct way to host APIs today :)

Cool, then you are free to spend time on it but do not expect us to research into that space for AdonisJS.

**Reply by LudovicLemarinel** [reply]
Would you approve a community provided package that would allow to run Adonis on serverless architectures ? 
I'm willing to spend my time on it...

**Reply by RomainLanz** [reply]
Yes, of course. Feel free to create whatever package you want! We will not stop you from trying. 😄

**Comment by thetutlage** [comment]
Yeah really not something that belongs here. We don't have bandwidth to research into the new space.

**Comment by McSneaky** [comment]
@tzsk something you can look for and use is to use only AdonisJS http server
It should work quite like described in readme usage section https://github.com/adonisjs/http-server#usage it should work on some FaaS platforms
If already digging in it might be worth to look at https://www.serverless.com/ too. Perhaps can integrate those 2 somehow

**Reply by tzsk** [reply]
Yup. You are exactly right. I did dig around a bit. There are references of Node Specific `http` & `https` references in `adonisjs/http-server` like you mentioned and also in the `adonisjs/core -> src/Ignitor/HttpServer`

The NODE runtime is tightly coupled with the framework, If we'd want to use it we will have to maintain a forked version of the framework. If we could extract Node Runtime objects Hypothetically, and accept the Things required from the node runtime like 'https' to a different object and accept the implementation on the bootstrap of `Ignitor.start()`

Then we can write an express adapter to run in AWS Lambda. And yes, we can cache the Bootstrapped Application Instance in memory inside a lambda. But a new replicated lambda instance will create a new App instance depending on load.

Also, Lambda warmer will take care of the bootstrapping and caching App Instance in memory.

**Comment by patie** [comment]
I try reopen this discussion, yeah its not best practice to have application in lambda, BUT, its possible and devs using this options daily. Great solution for MVP for example, you want check if product is valuable and not working on microservices and complex architecture from start. So why not API for SPA based on AdonisJs in Lambda.. You can check Laravel Vapor https://vapor.laravel.com, where people using Lambda for PHP with whole monolithic apps on serverless without issues. Its easy solution when people can deploy to serverless in few minutes.

**Reply by RomainLanz** [reply]
You took the wrong example.

PHP is a great candidate for FaaS because your application is not a long-running process.
The code is booted every-time you hit any routes.

**Reply by tzsk** [reply]
Then how is Express adapter running on a lambda?

Rather than reinventing the entire server from scratch if Adonis would have used Express underneath. We could have replaced that with any server adapter.

Like serverless-express or fastify.

I'm guessing Adonis doesn't support fastify either.

**Reply by RomainLanz** [reply]
I recommend you to re-read [my reply to your comment](https://github.com/adonisjs/core/discussions/1256#discussioncomment-38784).

It's not because some people are using the wrong tool that everyone must follow and adopt the wrong tool.

**Reply by thetutlage** [reply]
> Then how is Express adapter running on a lambda? 

@tzsk Maybe you can enlighten us with this? Since, you understand serveless better, it will be nice to know from you, how express is able to decouple itself from the Node http module

**Reply by thetutlage** [reply]
> I'm guessing Adonis doesn't support fastify either

Express and Fastify are not protocols, that we have to support them. It is like saying `"Rails doesn't support Sinatra"` and `"Django doesn't support Flask"` and `"Laravel doesn't support Silex"`

**Reply by thetutlage** [reply]
@patie Maybe you can create it for all of us and consider it as a great opportunity to contribute to the ecosystem?

**Reply by tzsk** [reply]
> > I'm guessing Adonis doesn't support fastify either
> 
> Express and Fastify are not protocols, that we have to support them. It is like saying `"Rails doesn't support Sinatra"` and `"Django doesn't support Flask"` and `"Laravel doesn't support Silex"`

Well I'm not saying that it is a protocol. Maybe it came off wrong. Express and Fastify are both implementation of Node Http Server. But  one can create express instance from Serverless Event and Context. (Already in place i.e. Serverless-Express package). And if any framework is built on top of express then they can easily accept that instance.

That way the framework gets the ability to run anywhere.

Express is just a bare-bone Server not a structured Framework in my opinion. And fastify is a bit faster but similar concept.

**Reply by thetutlage** [reply]
If Express and Fastify can run from the Serverless Event and Context, then Adonis can do (with required modifications).

The concern is not from someone proposing to add support for Severless. But concern is someone making statements like "Adonis doesn't support Fastify or Express and that is why it is not able to run in serverless environment."

**Reply by patie** [reply]
> You took the wrong example.
> 
> PHP is a great candidate for FaaS because your application is not a long-running process.
> The code is booted every-time you hit any routes.

@RomainLanz ok, this make sense, im PHP developer, newcomer in NodeJS, thanks for explanation.

**Reply by patie** [reply]
> @patie Maybe you can create it for all of us and consider it as a great opportunity to contribute to the ecosystem?

@thetutlage no need to be offensive, I didn't realize that NodeJS is a long running process, not like PHP. Im not skilled in NodeJS, so i cant help yet. See my above comment  https://github.com/adonisjs/core/discussions/1256#discussioncomment-386186

**Reply by tzsk** [reply]
> If Express and Fastify can run from the Serverless Event and Context, then Adonis can do (with required modifications).
> 
> The concern is not from someone proposing to add support for Severless. But concern is someone making statements like "Adonis doesn't support Fastify or Express and that is why it is not able to run in serverless environment."

No no. God no. I certainly didn't mean to say it like that. 
I'm sure we can modify the internals of Adonis to accept Serverless Event and Context instead of booting Node HTTP module and make our own adapter that we would have to maintain.
It all boils down to the platform. If we try and make Adonis on top of Express then we won't need to manage our own Serverless adapter. AWS will manage that for us relative to the changes in Serverless ecosystem going forward.

https://www.npmjs.com/package/aws-serverless-express

**Reply by thetutlage** [reply]
@tzsk Cool. Maybe it was a misunderstanding on my end. Thanks for the clarification 🙂

**Reply by McSneaky** [reply]
Slightly related topic
I tried to hack AdonisJS work on both Deno and ES4X
Got the ES4X version somewhat working with nice performance boost, but I had to disable tons of stuff (cookies, sessions, crypto etc). Just simply coz I didn't have time to add "proxy" layer between all AdonisJS Node calls and ES4X things (like fs, http, net etc differ a lot)
ES4X provides way to overwrite certain packages. For example load `@adonis/lucid` from AdonisJS, but overwrite `http-server` with my custom implementation (which is mostly copy-paste of original, with some modifications to build in lib calls)

Maybe.. Just maybe something like this can be done in the serverless world. If anyone is more familiar with it, then making AdonisJS `http-server` work on serverless would be good step forward

**Reply by douglasgc** [reply]
@RomainLanz you are not correct in the statement "Booting-up the application everytime you hit your FaaS URI is extremely inefficient.".
There is a way to maintain the context, and to have extremely efficient requests. I showed you how to do it in my answer.

**Comment by douglasgc** [comment]
I solved it this way:

1. Install package `npm i --save aws-lambda-create-request-response`
2. Create a file `aws-lambda.js` with the content:
```javascript
const { Ignitor } = require('@adonisjs/ignitor')
const createRequestResponse = require('aws-lambda-create-request-response')
const http = require('http')

var app;

function handler(event, context, callback) {
    context.callbackWaitsForEmptyEventLoop =
        process.env.WAIT_FOR_EMPTY_EVENT_LOOP === 'yes'
    const { req, res } = createRequestResponse(event, callback)
    app(req, res)
}

function bootstrapServer() {
    return new Ignitor(require('@adonisjs/fold'))
        .appRoot(__dirname)
        .fireHttpServer((handler) => {
            app = handler;
            return http.createServer(handler);
        })
        .catch(console.error);
}
exports.proxy = (event, context, callback) => {
    if (app == undefined) {
        bootstrapServer().then(() => handler(event, context, callback))
        return;
    };
    handler(event, context, callback);
}
```

3. Create a file serverless.yml
```yml
service: service-users

provider:
  name: aws
  runtime: nodejs12.x
  stage: qualidade
  region: sa-east-1

functions:
  index:
    handler: aws-lambda.proxy
    events:
      - http:
          cors: true
          path: '/'
          method: any
      - http:
          cors: true
          path: '{proxy+}'
          method: any
```
          
4. run `sls deploy`

**Reply by douglasgc** [reply]
The performance I found was:
first request (code start): 400 ms
response to requests from then on: 50ms

**Reply by popovicieduard** [reply]
You have a typo after `method: any` additional "`"

**Reply by douglasgc** [reply]
thanks @popovicieduard. fixed

**Reply by popovicieduard** [reply]
Did you try to run this serverless adapter for v5? I tried adapting your code to v5 but I get a error regarding the .adonisrc.json not being present in the root of the folder.

**Reply by douglasgc** [reply]
No, i just try in v4. 
where did the error happen? I will take time to fix this problem: D

**Reply by popovicieduard** [reply]
when running sls deploy or sls offline, I tracked the issue to adonis/application/Application.js's method:
```
    loadRcFile() {
        return this.resolveModule('./.adonisrc.json', () => {
            throw new Error('AdonisJS expects ".adonisrc.json" file to exist in the application root');
        });
    }
```

**Reply by popovicieduard** [reply]
Also, my 2 files:

lambda.ts
```ts
import 'reflect-metadata'
import { Ignitor } from '@adonisjs/core/build/standalone'
import createRequestResponse from 'aws-lambda-create-request-response'
//@ts-ignore
import http from 'http'

var app: any

function handler(event: any, context: any, callback: any) {
  context.callbackWaitsForEmptyEventLoop =
    process.env.WAIT_FOR_EMPTY_EVENT_LOOP === 'yes'
  const { req, res } = createRequestResponse(event, callback)
  app(req, res)
}

function bootstrapServer() {
  return new Ignitor(__dirname)
    .httpServer()
    .start(handler => {
      app = handler
      return http.createServer(handler)
    })
}

export const handle = (event:any, context:any, callback:any) => {
  try {
    if (app == undefined) {
      bootstrapServer()
      return;
    };
    handler(event, context, callback);
  } catch (error) {
    console.error(error)
  }
}
```

serverless.yml:
```yml
org: upnance
app: serverless-adonis
service: adonisv5

provider:
  name: aws
  runtime: nodejs12.x
  environment:
    NODE_ENV: dev

plugins:
  - serverless-plugin-typescript
  - serverless-offline

functions:
  index:
    handler: lambda.handle
    events:
      - http:
          cors: true
          path: '/'
          method: any
      - http:
          cors: true
          path: '{proxy+}'
          method: any

```

**Reply by douglasgc** [reply]
I didn't understand, did you fix the problem or not?  :D

**Reply by popovicieduard** [reply]
I did not, I pasted my configuration to reproduce this.
I just found out where the error is coming from.

**Reply by tomhatzer** [reply]
@popovicieduard can you possibly share the solution for your error if you've found it like you said in your last sentence? :) Thank you very much!

**Reply by popovicieduard** [reply]
@tomhatzer I did not find a solution, just pointed out that there is an issue with the .adonisrc.json not being loaded: `Error('AdonisJS expects ".adonisrc.json" file to exist in the application root');`

**Reply by tomhatzer** [reply]
@popovicieduard oh ok, then I misinterpreted your last sentence. Thanks for finding the source of the error. I'll also try to get it working and I'll keep you updated here if I find a solution for it.

**Reply by popovicieduard** [reply]
@tomhatzer thank you, best of luck

**Reply by tuliovargas** [reply]
Hi @douglasgc , I liked your solution! Can we get in touch in order to get some help from you?

**Reply by douglasgc** [reply]
@tuliovargas yes, send-me email 'douglasdm97@gmail.com'

**Reply by tomhatzer** [reply]
I'd also be interested in a solution. I've been able to get around the `adonisrc.json` error but I'm unable to get the request/response stuff to work as I'm not that experienced with Typescript.

**Reply by popovicieduard** [reply]
@tomhatzer Do you mind sharing a repo ? I am also interested in making this work.

**Reply by popovicieduard** [reply]
@tomhatzer @douglasgc 
I managed to make this work but it's not really having the expected result.

```
org: app
app: serverless-adonis
service: adonisv5

provider:
  name: aws
  runtime: nodejs14.x
  environment:
    NODE_ENV: development
    PORT: 3333
    HOST: localhost
    APP_KEY: TVTj8OCKxhHhHWeyZHkVhK1TPL7ghIK9

plugins:
  - serverless-plugin-typescript
  - serverless-offline

functions:
  index:
    handler: lambda.handle
    events:
      - http:
          cors: true
          path: '/'
          method: any
      - http:
          cors: true
          path: '{handle+}'
          method: any
```

lambda.ts
```
import 'reflect-metadata'
import { Ignitor } from '@adonisjs/core/build/standalone'
import createRequestResponse from 'aws-lambda-create-request-response'

var app: any

function handler(event: any, context: any, callback: any) {
  context.callbackWaitsForEmptyEventLoop = process.env.WAIT_FOR_EMPTY_EVENT_LOOP === 'yes'
  const { req, res } = createRequestResponse(event, callback)
  app(req, res)
}

function createServer() {
  return new Ignitor(__dirname).httpServer().start().catch(console.error)
}

exports.handle = (event: any, context: any, callback: any) => {
  try {
    if (app === undefined) {
      app = createServer()
    }
    handler(event, context, callback)
  } catch (error) {
    console.error(error)
  }
}

```

tsconfig
```
{
  "include": [
    "**/*"
  ],
  "exclude": [
    "node_modules",
    "build"
  ],
  "compilerOptions": {
    "baseUrl": ".",
    "preserveConstEnums": true,
    "strictNullChecks": true,
    "target": "ESNext",
    "moduleResolution": "node",
    "module": "commonjs",
    "allowSyntheticDefaultImports": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "esModuleInterop": true,
    "noUnusedLocals": true,
    "skipLibCheck": true,
    "noUnusedParameters": true,
    "removeComments": true,
    "declaration": false,
    "lib": [
      "ESNext"
    ],
    "outDir": ".build",
    "rootDir": "./",
    "sourceMap": true,
    "paths": {
      "App/*": [
        "./app/*"
      ],
      "Config/*": [
        "./config/*"
      ],
      "Contracts/*": [
        "./contracts/*"
      ],
      "Database/*": [
        "./database/*"
      ]
    },
    "types": [
      "@adonisjs/core",
      "@adonisjs/repl"
    ]
  }
}
```

Keep in mind that you have to run `npm run build` and `sls offline` afterwards,

If anyone has any idea on how can I solve these 2 issues it would be great.

1st Issue: I can't run the lambda.ts directly, I need to compile the files from ts with `npm run build` and I don't know how this would influence the process of running this in prod.

2nd issue: I can't wrap my head around how can I make the lamba file to properly work, right now the current behaviour is like this: When hitting the `http://localhost:3000/dev` the server will start and afterwards `http://localhost:3333/` will be available to handle requests, what is the desired behaviour is when hitting `http://localhost:3000/dev` it would start the server if not started yet and then handle the request.

**Reply by tomhatzer** [reply]
@popovicieduard

For your first issue, have a look at the `serverless-plugin-typescript` plugin. This automatically compiles your typescript to normal javascript while deploying.

The second issue has a basic "problem": You don't want to spin up another server as you'll already have one running on port 3000.
What I've tried to get around this:
- vendia/serverless-express
- aws-lambda-http-server
- aws-lambda-create-request-response
- serverless-http

What we'd need to accomplish is to move the stuff that the server is running into aws lambda compatible requests and responses. We'd have to move the things from Ignitor out into a custom Lambda handler which takes care of incoming requests and outgoing responses. With this change we should be able to get it up and running without having to run a separate server on a second port.

**Reply by popovicieduard** [reply]
@tomhatzer wanna open up a repo ? I'm trying right now to solve this problem. (more heads the better)

**Reply by tomhatzer** [reply]
@popovicieduard I've created an empty repository here and invited you: https://github.com/tomhatzer/adonis-serverless
I've got my last test case on my local system, the other changes are sadly already gone... But knowledge is still here.
You can push your code if you want so we can work together on it.

**Reply by capopovici** [reply]
@tomhatzer  I think I found a relevant repo for this:
https://github.com/satheler/satheler-survey

**Reply by tomhatzer** [reply]
@capopovici the package this repos is using is this one here: https://github.com/satheler/s12r
Looks very promising and also has an example to try.

**Reply by capopovici** [reply]
Hmm, trying to run the example but I have 2 errors regarding ioc:
`E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Core/Server" namespace from the IoC Container`
`E_IOC_LOOKUP_FAILED: Cannot resolve "Adonis/Core/Route" namespace from the IoC Container`

**Reply by tomhatzer** [reply]
Adonis/Core/Server is now here: https://github.com/adonisjs/http-server/blob/develop/src/Server/index.ts
Adonis/Core/Route is now here: https://github.com/adonisjs/http-server/blob/develop/src/Router/Route.ts

Maybe we can ask @satheler if he can make this compatible with the latest version if we can't get it to work. 🤔

**Reply by capopovici** [reply]
@tomhatzer finallyyyy, solved the issues, will push in a bit

**Comment by capopovici** [comment]
For anyone looking at this.

https://github.com/tomhatzer/adonis-serverless

`npm install`
`npm run serverless:start`

**Reply by tomhatzer** [reply]
Woah this is huge! Awesome work @capopovici !! I'll test this today evening too and report anything that may occur that shouldn't occur.

**Reply by thetutlage** [reply]
Great work @capopovici. I myself is not much into serverless, but glad to see community pushing limits 🙂

**Reply by capopovici** [reply]
Neither am I but was interesting and fun to make the framework work with serverless.

**Reply by tomhatzer** [reply]
@capopovici i‘ll put a readme up with general information and that this is coming from you.

**Reply by tomhatzer** [reply]
@capopovici I've added a basic README.md file here: https://github.com/tomhatzer/adonis-serverless/blob/master/README.md

Can you please check if this is ok for you or otherwise make the necessary changes?

**Comment by aonesur** [comment]
The NODE runtime is tightly coupled with the framework, If we'd want to use it we will have to maintain a forked version of the framework. [surveyzop](https://surveyzop.com/)

**Comment by buzzware** [comment]
I've had success for years running node, express and Feathers.js apps on a single Google Firebase/Cloud Function. They also run on node on my local machine.
Google officially supports express on Cloud Functions.

https://github.com/buzzware/function-split-test

1) Is AdonisJs fundamentally different to express such that this won't be so easy or efficient to run Adonis on Cloud Functions?
2) Is AWS Lambda fundamentally different to Cloud Functions for running something like express ?

**Reply by davidharting** [reply]
You can run Node.js in AWS lambda serverless functions. I think it would be possible to write an adapter for Adonis applications to work on lambda and other serverless environments. 

There could be concerns about cold starts but I think it would be interesting and could be a worthwhile tradeoff for some applications.

**Comment by pitchinnate** [comment]
Spent some time trying to get this to work and I believe I have a solution at least for Adonis Commands running in a docker image on AWS Lambda, I don't think this would be a good idea for serving HTTP requests, more for running one off functions. I plan on using it with SQS triggering Lambda functions to have it process queued jobs.

I made a small repo that has some instructions and files to make it easier for others to give it a try. Hopefully it helps some other people out.

https://github.com/pitchinnate/AdonisLambda
